src/clients/googleAnalytics.ts

```diff
@@ -0,0 +1,27 @@
+type EventParameters = Parameters<(eventName: string, parameters?: Record<string, string>) => void>;
+
+const googleAnalytics = () => {
+  let gtag: Gtag.Gtag;
+  const loadGTag = (callback: (...args: EventParameters) => void) => (...args: EventParameters) => {
+    gtag = window?.gtag;
+    if (!gtag && process.env.NODE_ENV !== 'production') {
+      console.warn('gtag not loaded. Google Analytics event not sent', args);
+    } else {
+      callback(...args);
+    }
+  };
+
+  const sendEvent = loadGTag((eventName: string, parameters?: Record<string, string>) => {
+    gtag('event', eventName, parameters);
+  });
+
+  const buttonPressed = (buttonName: string, parameters: Record<string, string> = {}) => {
+    sendEvent('button_pressed', { button_name: buttonName, ...parameters });
+  };
+
+  return {
+    buttonPressed,
+  };
```

Comment:
That looks very good, only exposing specific events is great in my opinion.

As the project grows and we add more events, we will have more and more functions that look like `buttonPressed`, some of which will share the exact same type definition apart from the name of the event being used.
In that case, we could instead expose the `sendEvent` function directly, but use the type definition to restrict its usage to only the events we want to use:

```typescript
// Define the types of each event and its parameters
export type AnalyticEventMap = {
  button_pressed: {
    button_name: string;
  };
};

// Extract the event name from the map
export type AnalyticEventName = keyof AnalyticEventMap;

const googleAnalytics = {
  // ...

  sendEvent: function<E extends AnalyticEventName>(
    eventName: E,
    parameters: AnalyticEventMap[E] extends undefined
      ? never
      : AnalyticEventMap[E]
  ) {
    gtag('event', eventName, parameters);
  }
};

export default googleAnalytics;
```

What do you think?

Reply:
Yeah, I like it